csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -698,31 +696,94 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
-            }
-            finally
+        /// <summary>
+        ///  This method return a collection of DisposableNamedOnnxValue as in other interfaces
+        ///  Query names from OrtIoBinding object and pair then with the array of OrtValues returned
+        /// from OrtIoBinding.GetOutputValues()
+        /// 
+        /// </summary>
+        /// <param name="runOptions">RunOptions</param>
+        /// <param name="ioBinding">OrtIoBinding instance with bindings</param>
+        /// <param name="names">optional parameter. If you already know the names of the outputs you can save a native
+        /// call to retrieve output names. They will be paired with the returned OrtValues and combined into DisposbleNamedOnnxValues.
+        /// Otherwise, the method will retrieve output names from the OrtIoBinding instance.
+        /// It is an error if you supply a different number of names than the returned outputs</param>
+        public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> RunWithBindingAndNames(RunOptions runOptions, OrtIoBinding ioBinding, string[] names = null)
+        {
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtRunWithBinding(Handle, runOptions.Handle, ioBinding.Handle));
+            using (var ortValues = ioBinding.GetOutputValues())
             {
-                if (nameHandle != IntPtr.Zero)
+                string[] outputNames = names;
+                if (outputNames == null)
+                {
+                    outputNames = ioBinding.GetOutputNames();
+                }
+
+                if (outputNames.Length != ortValues.Count)
+                {
+                    throw new OnnxRuntimeException(ErrorCode.InvalidArgument,
+                        "Number of specified names: " + names.Length + " does not match the output number: " +
+                        ortValues.Count);
+                }
+
+                var result = new DisposableList<DisposableNamedOnnxValue>(outputNames.Length);
+                try
+                {
+                    for (int i = 0; i < outputNames.Length; ++i)
+                    {
+                        var ortValue = ortValues.ElementAt(i);
+                        result.Add(DisposableNamedOnnxValue.CreateTensorFromOnnxValue(outputNames[i], ortValue.Handle));
+                        // We transferred ownership of the handle.
+                        // Make sure it is not disposed here
+                        ortValue.Disown();
+                    }
+                } catch(Exception e)
                 {
-                    NativeMemoryAllocator.DefaultInstance.FreeMemory(nameHandle);
+                    result.Dispose();
+                    throw e;
                 }
+                return result;
             }
+        }
 
-            return str;
+        /// <summary>
+        /// Ends profiling for the session. Returns the profile file name.
+        /// 
+        public string EndProfiling()
```

Comment:
wasn't there an EndProfiling() already exposed ?

Reply:
I clean up the function since it was simple.

---
In reply to: [463238636](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r463238636) [](ancestors = 463238636)